lib/http/request.ts

```diff
@@ -142,8 +158,74 @@ export function httpRequest(sdk: OktaAuthHttpInterface, options: RequestOptions)
     withCredentials
   };
 
-  var err, res;
-  return sdk.options.httpRequestClient!(method!, url!, ajaxOptions)
+  var err, res, promise;
+
+  if (isIOS()) {
+    let waitForVisibleAndAwakenDocument: () => Promise<void>;
+    let waitForAwakenDocument: () => Promise<void>;
+    let recursiveFetch: () => Promise<HttpResponse>;
+
+    // Safari on iOS has bug:
+    //  Performing `fetch` right after document became visible can fail with `Load failed` error.
+    // Running fetch after short timeout fixes this issue.
+    waitForAwakenDocument = () => {
+      const timeSinceDocumentIsVisible = Date.now() - dateDocumentBecameVisible;
+      if (isIOS() && timeSinceDocumentIsVisible < IOS_PAGE_AWAKEN_TIMEOUT) {
+        return new Promise<void>((resolve) => setTimeout(() => {
+          if (!document.hidden) {
+            resolve();
+          } else {
+            resolve(waitForVisibleAndAwakenDocument());
+          }
+        }, IOS_PAGE_AWAKEN_TIMEOUT - timeSinceDocumentIsVisible));
+      } else {
+        return Promise.resolve();
+      }
+    };
+
+    // Returns a promise that resolves when document is visible for 500 ms
+    waitForVisibleAndAwakenDocument = () => {
+      if (document.hidden) {
+        let pageVisibilityHandler: () => void;
+        return new Promise<void>((resolve) => {
+          pageVisibilityHandler = () => {
+            if (!document.hidden) {
+              document.removeEventListener('visibilitychange', pageVisibilityHandler);
+              resolve(waitForAwakenDocument());
+            }
+          };
+          document.addEventListener('visibilitychange', pageVisibilityHandler);
+        });
+      } else {
+        return waitForAwakenDocument();
+      }
+    };
+
+    // Restarts fetch on 'Load failed' error
+    // This error can occur when `fetch` does not respond
+    //  (due to CORS error, non-existing host, or network error)
```

Comment:
Is there a way to differentiate between/among these types of errors?

Seems to me `isNetworkError` will be true in all those cases and retry the fetch, although **we don't want to retry** on CORS, non-existent host, etc.

LMK if I'm misreading/misunderstanding the behavior

Reply:
- Google search results for "[safari types of network error](https://www.google.com/search?q=safari+types+of+network+error&rlz=1C5GCCM_en&oq=safari+types+of+network+error&gs_lcrp=EgZjaHJvbWUyBggAEEUYOdIBCDcxNjdqMGo3qAIAsAIA&sourceid=chrome&ie=UTF-8)"
- Google search results for "[chrome devtools types of network error](https://www.google.com/search?q=chrome+devtools+types+of+network+error&sca_esv=31fe99a6d4a2597d&rlz=1C5GCCM_en&ei=IHJkZ_-FGvSv5NoPoK3TmA4&ved=0ahUKEwj_lb3zxrSKAxX0F1kFHaDWFOMQ4dUDCBA&uact=5&oq=chrome+devtools+types+of+network+error&gs_lp=Egxnd3Mtd2l6LXNlcnAiJmNocm9tZSBkZXZ0b29scyB0eXBlcyBvZiBuZXR3b3JrIGVycm9yMgoQABiwAxjWBBhHMgoQABiwAxjWBBhHMgoQABiwAxjWBBhHMgoQABiwAxjWBBhHMgoQABiwAxjWBBhHMgoQABiwAxjWBBhHMgoQABiwAxjWBBhHSNAHUI4HWI4HcAF4AZABAJgBAKABAKoBALgBA8gBAPgBAZgCAaACBpgDAIgGAZAGB5IHATGgBwA&sclient=gws-wiz-serp)"

Reply:
Probably good to add some tests to ensure we don't catch and retry CORS errors, etc

Reply:
You're correct, `isNetworkError` will be true in all those 3 cases. 
There is no way to distinguish programmatically in JS. 
(The difference can be seen only in browser console)
That's why I added `canRetry` option to enable retry logic only for polling requests.
Polling requests are repeatable by design so I think it's fine to retry it due to real network loss.

Since all API requests should be performed to the same host and poll request is never a first request in a list of ones performed by SIW, possible CORS error should be caught before polling is started.

Reply:
Thanks. Is there any logic to say we should stop trying polling after n failed attempts?